src/bitcoin-cli.cpp

```diff
@@ -485,56 +500,58 @@ static int CommandLineRPC(int argc, char *argv[])
             method = args[0];
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
-
-        // Execute and handle connection failures with -rpcwait
-        const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
-        do {
-            try {
-                const UniValue reply = CallRPC(rh.get(), method, args);
-
-                // Parse reply
-                const UniValue& result = find_value(reply, "result");
-                const UniValue& error  = find_value(reply, "error");
-
-                if (!error.isNull()) {
-                    // Error
-                    int code = error["code"].get_int();
-                    if (fWait && code == RPC_IN_WARMUP)
-                        throw CConnectionFailed("server in warmup");
-                    strPrint = "error: " + error.write();
-                    nRet = abs(code);
-                    if (error.isObject())
-                    {
-                        UniValue errCode = find_value(error, "code");
-                        UniValue errMsg  = find_value(error, "message");
-                        strPrint = errCode.isNull() ? "" : "error code: "+errCode.getValStr()+"\n";
-
-                        if (errMsg.isStr())
-                            strPrint += "error message:\n"+errMsg.get_str();
-
-                        if (errCode.isNum() && errCode.get_int() == RPC_WALLET_NOT_SPECIFIED) {
-                            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+        UniValue reply = ConnectAndCallRPC(rh.get(), method, args);
+
+        // Parse reply
+        UniValue result = find_value(reply, "result");
+        const UniValue& error = find_value(reply, "error");
+        if (!error.isNull()) {
+            // Error
+            int code = error["code"].get_int();
+            if (code == RPC_IN_WARMUP && gArgs.GetBoolArg("-rpcwait", false)) {
+                throw CConnectionFailed("server in warmup");
```

Comment:
After commit _cli: extract connection try/wait/failure logic_, this throw no longer gets caught.

Reply:
Thanks @jnewbery, good catch on the throw! Updated that commit as per
`git diff e9a271b ed06899`
```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 53c6f3a655..18317574ad 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -419,6 +420,13 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler *rh, const std::string& str
    const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
     do {
         try {
             response = CallRPC(rh, strMethod, args, rpcwallet);
+
+            if (fWait) {
+                const UniValue& error = find_value(response, "error");
+                if (!error.isNull() && error["code"].get_int() == RPC_IN_WARMUP) {
+                    throw CConnectionFailed("server in warmup");
+                }
+            }
             break; // Connection succeeded, no need to retry.
         }
         catch (const CConnectionFailed&) {
@@ -507,12 +514,8 @@ static int CommandLineRPC(int argc, char *argv[])
         const UniValue& error = find_value(reply, "error");
         if (!error.isNull()) {
             // Error
-            int code = error["code"].get_int();
-            if (code == RPC_IN_WARMUP && gArgs.GetBoolArg("-rpcwait", false)) {
-                throw CConnectionFailed("server in warmup");
-            }
             strPrint = "error: " + error.write();
-            nRet = abs(code);
+            nRet = abs(error["code"].get_int());
             if (error.isObject()) {
                 UniValue errCode = find_value(error, "code");
                 UniValue errMsg  = find_value(error, "message");
```